### ISSUE TITLE
New context API has delayed update behavior

### ISSUE THREAD

**Issue by LPGhatguy** [Roblox/roact#259]
If I update a context value in the same update as adding a consumer of that value, the consumer will receive the old update. I'm still working on a better standalone repro, but basically:

```lua
local tree = Roact.mount(
	Roact.createElement(Foo.Provider, {
		value = 1,
	})
)

Roact.update(
	tree,
	Roact.createElement(Foo.Provider, {
		value = 2,
	}, {
		Consumer = Roact.createElement(Foo.Consumer, {
			render = function(value)
				print("value", value)
			end
		}),
	})
)
```

This code should print `value 2`, but I believe (once I pull a repro out) that it would instead print `value 1`.

This manifested itself as a bug in a game that I'm working on.

**Comment by ZoteTheMighty** [comment]
I wonder if this is somehow the culprit: https://github.com/Roblox/roact/blob/f7d2f1ce1dd69120f457c2f8032bb184eb8f0c29/src/createContext.lua#L43-L47

The callback for the binding subscription is using `getValue`, which certainly _shouldn't_ be returning an old value (we set the value before we fire the change signal), but maybe somehow is?. This could be a bug with bindings as well, though in this case I think we should also be using the more idiomatic binding subscription logic like we do here: https://github.com/Roblox/roact/blob/f7d2f1ce1dd69120f457c2f8032bb184eb8f0c29/src/RobloxRenderer.lua#L73-L94

**Comment by LPGhatguy** [comment]
Do you think this is because of deferred `setState`? Mid-reconcile, we'll update the value of the binding, which will call `setState`, but then we continue with the rest of the render with the old value. Only after the update would we then come back and perform the result of `setState`!
